jaeger-core/src/main/java/io/jaegertracing/internal/clock/CurrentTimeSupplier.java

```diff
@@ -0,0 +1,103 @@
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.jaegertracing.internal.clock;
+
+import java.lang.reflect.Method;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import lombok.val;
+
+/**
+ * This interface contains is responsible for supplier the current time in microseconds
+ * to the {@link SystemClock} implementation. It contains two implementations
+ * <ul>
+ *     <li>{@link MillisAccuracy} - For when the JDK version is below 9. Calls
+ *     {@link System#currentTimeMillis} and multiplies the result
+ *     </li>
+ *     <li>{@link MicrosAccuracy} - For when the JDK version is equal to or above 9.
+ *     It loads and uses java.time types Instant, ChronoUnit and Temporal (available since JDK 1.8)
+ *     via reflection in order generate real microsecond accurate timestamp.
+ *     </li>
+ * </ul>
+ * <p>
+ *     NOTE: the {@link MicrosAccuracy} implementation pre-generates and stores all reflective accessors on
+ *     class-loading, minimizing reflective operations in real-time to just 2 non-parameterized method calls.
+ *     This class is also further optimized by the JIT compiler and Hot-Spot JVM upon recurring use, so we should
+ *     not suffer any notable performance penalties here.
+ * <p>
+ * <p>
+ *     While it would have been preferable to generate multi-version/release jars to differentiate between the two cases
+ *     The project currently targets JDK 1.6, and reworking gradle and release cycle to accommodate for this would be
+ *     too much big of change for now.
+ * <p>
+ * <p>
+ *     Thus, the current implementation remains our best option
+ *
+ * @author <a href="mailto:ishinberg0@gmail.com">Idan Sheinberg</a>
+ * @see System#currentTimeMillis()
+ */
+interface CurrentTimeSupplier {
+
+  long micros();
+
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  final class MillisAccuracy implements CurrentTimeSupplier {
+
+    static final CurrentTimeSupplier INSTANCE = new MillisAccuracy();
+
+    @Override
+    public long micros() {
+      return System.currentTimeMillis() * 1000;
+    }
+  }
+
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  final class MicrosAccuracy implements CurrentTimeSupplier {
+
+    static final CurrentTimeSupplier INSTANCE = new MicrosAccuracy();
+
+    private static final Method NOW;
+    private static final Object EPOCH;
+    private static final Object CHRONO_UNIT_MICROS;
+    private static final Method CHRONO_UNIT_BETWEEN;
+
+    static {
+      try {
+        val classLoader = ClassLoader.getSystemClassLoader();
```

Comment:
Can we use FQCN instead of reflection? Maybe it's better to move this all together to a separate file and load only if the instant exists or the java 9 is present. 

It seems redundant to be using reflection for the instant class and also `java.version` can we unify it and use only one?

Reply:
@pavolloffay I agree, it would  have been much better to go multi release.
However, it'll require additional changes to code layout and gradle definitions. and might create development inconvenience for IDE users (depends on the IDE and it's understanding of  gradle `sourceSets`/`configurations`.

Again this is not my codebase - I'll need a strong clear decision on what's acceptable as part of the scope of this PR.

let me know.

Reply:
We don't want to do multi-release nor mess up with grade config.

My suggestion is to just move the jdk9 code to a separate class. The client code would load the class only when running on JDK9 - checked by `java.version` or the presence of JDK9 classes. This should work well as the class would be loaded only when referenced.

Reply:
So just to clarify - keep the code as it is (reflection based access to newer classes ) , but separate different use cases to different classes?

Reply:
yes

Reply:
Fixed